### ISSUE TITLE
AuthSession.useAuthRequest never loads request due to rejected promise from getRandomBytesAsync

### ISSUE THREAD

**Issue by jfbn** [expo/expo#9974]
### Summary of Issue 
I am trying to implement social logins (starting with Facebook) by making use of [AuthSession](https://docs.expo.io/versions/latest/sdk/auth-session/).
According to the [documentation](https://docs.expo.io/versions/latest/sdk/auth-session/), it is supposed to return a AuthRequest object once it has been loaded.
This never happens due to a unhandled rejected promise with this error:
`TypeError: null is not an object (evaluating '_ExpoRandom.default.getRandomBytesAsync')`

### Environment - output of `expo diagnostics` & the platform(s) you're targeting
System:
      OS: Linux 5.4 Ubuntu 20.04.1 LTS (Focal Fossa)
      Shell: 5.0.17 - /bin/bash
    Binaries:
      Node: 14.4.0 - ~/.nvm/versions/node/v14.4.0/bin/node
      npm: 6.14.7 - ~/.nvm/versions/node/v14.4.0/bin/npm
    npmPackages:
      expo: ~38.0.8 => 38.0.8 
      react: ~16.11.0 => 16.11.0 
      react-dom: ~16.11.0 => 16.11.0 
      react-native: ~0.62.2 => 0.62.2 
      react-native-web: ~0.11.7 => 0.11.7 
      react-navigation: ^4.4.0 => 4.4.0 
    npmGlobalPackages:
      expo-cli: 3.24.2
 
### Steps to Reproduce

1. Initialize a new bare expo project with `expo init`
2. Install dependencies: `expo install expo-auth-session`
3. Call useAuthRequest function like so:
`const [request, response, promptAsync] = useAuthRequest(
    {
      clientId: "<my Facebook app id>",
      scopes: ["public_profile", "user_likes"],
      redirectUri: makeRedirectUri({
        native: "fb<my Facebook app id>://authorize",
      }),
    },
    {
      authorizationEndpoint: "https://www.facebook.com/v6.0/dialog/oauth",
      tokenEndpoint: "https://graph.facebook.com/v6.0/oauth/access_token",
    }
  );`

### Expected Behavior vs Actual Behavior
I expected useAuthRequest to eventually load the request object so I can prompt the user for access.

What happens instead is that useAuthRequest never succesfully loads the request. It appears that some internal part of the function relies on [Expo random getBytesAsync()](https://docs.expo.io/versions/latest/sdk/random/#randomgetrandombytesasyncbytecount) and for some reason that is `null` when used inside useAuthRequest.

Note: I have no issues manually using the `getRandomBytesAsync` function manually in an isolated environment.

This is my first Github issue so forgive me if I am forgetting something. I will keep a close eye here and update according to eventual comments.

**Comment by cruzach** [comment]
hey @jfbn, is this on iOS? Just wanted to double-check that you ran `pod install` after installing `expo-auth-session`

**Comment by jfbn** [comment]
Hi @cruzach, thanks for going over this.

I have only been testing on Android. I have not been able to reproduce the error on a snack. I know they run Expo 38.0.0 and I am on 38.0.8, that's the only discrepancy I have been able to identify so far.

**Comment by yujuiting** [comment]
I am not an expert about expo.
Seems it's caused by local `expo-auth-session`'s dependency.
I work around it by `rm -rf node_modules/expo-auth-session/node_modules/expo-random`.
And add this script into post-install command.
Hope it's helpful for you.

**Comment by jfbn** [comment]
> I am a newbie about expo.
> Seems it's caused by local `expo-auth-session`'s dependency.
> I work around it by `rm -rf node_modules/expo-auth-session/node_modules/expo-random`.
> And add this script into post-install command.
> Hope it's helpful to you.

Thank you so much for your comment @yujuiting. Could you point me towards any resource that could explain how I add the rm script into my post-install command..?

**Comment by Joozty** [comment]
The same issue here 😢...  and I believe bare workflow label is wrong. it's actually managed workflow.

**Comment by jfbn** [comment]
I am on a the bare workflow.

**Comment by zwass** [comment]
I run into the same issue with a brand new app and the Strava instructions copied from the Expo docs.

Is the expo auth functionality entirely broken?

```jsx
import { StatusBar } from 'expo-status-bar';
import React from 'react';
import { StyleSheet, Text, View } from 'react-native';

import * as WebBrowser from 'expo-web-browser';
import { makeRedirectUri, useAuthRequest } from 'expo-auth-session';
import { Button } from 'react-native';

// Endpoint
const discovery = {
    authorizationEndpoint: 'https://www.strava.com/oauth/mobile/authorize',
    tokenEndpoint: 'https://www.strava.com/oauth/token',
    revocationEndpoint: 'https://www.strava.com/oauth/deauthorize',
};

export default function App() {
    const [request, response, promptAsync] = useAuthRequest(
        {
            clientId: '44444',
            scopes: ['activity:read_all'],
            // For usage in managed apps using the proxy
            redirectUri: makeRedirectUri({
                // For usage in bare and standalone
                // the "redirect" must match your "Authorization Callback Domain" in the Strava dev console.
                native: 'test://redirect',
            }),
        },
        discovery
    );

    React.useEffect(() => {
        if (response?.type === 'success') {
            const { code } = response.params;
        }
    }, [response]);

    return (
        <Button
            disabled={!request}
            title="Login"
            onPress={() => {
                promptAsync();
            }}
        />
    );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
    alignItems: 'center',
    justifyContent: 'center',
  },
});
```

**Comment by EvanBacon** [comment]
Seems this happened because the version of `expo-auth-session` I published earlier today was using `@next` versions of `expo-crypto` and `expo-random` which caused them to be installed in `node_modules/expo-auth-session/node_modules/*` instead of `node_modules/*` which is why the native module wasn't registered.

I just republished with the lower versions `yarn add expo-auth-session@~1.5.1` should solve the problem for now. 

Pro Tip: You can use "yarn resolutions" to force the version of `expo-random` to be the same across all packages. If you get blocked on an issue like this in the future this can help unblock you for a bit!

**Comment by brentvatne** [comment]
i republished `expo-random@8.3.1` to be compatible with sdk-38 and will publish what was formerly known as `expo-random@8.3.0` as `expo-random@9.0.0`

if you uninstall expo-auth-session and reinstall that should solve your issue in sdk-38

**Comment by zwass** [comment]
I get this now when running the project:

```
Some of your project's dependencies are not compatible with currently installed expo package version:
 - expo-auth-session - expected version range: ~1.4.0 - actual version installed: ~1.5.1
Your project may not work correctly until you install the correct versions of the packages.
```

Is there a resolution? Can it be safely ignored? If so, can it be silenced?

**Comment by zwass** [comment]
My above concern seems to be fixed by @brentvatne latest comment.

I did `yarn remove expo-auth-session && expo install expo-auth-session`.

**Comment by plaporte** [comment]
> My above concern seems to be fixed by @brentvatne latest comment.
> 
> I did `yarn remove expo-auth-session && expo install expo-auth-session`.

Ah thank you so much. This worked perfectly!

**Comment by lostchopstik** [comment]
I'm having the same error message as originally descripted and the command `yarn remove expo-auth-session && expo install expo-auth-session` did not work for me. Is there anything else I should try??

**Comment by santiripper** [comment]
Me too, i'm having the same problem on SDK 39.

> I'm having the same error message as originally described and the command `yarn remove expo-auth-session && expo install expo-auth-session` did not work for me. Is there anything else I should try??
> 
> Also, I am on SDK 39.

**Comment by santiripper** [comment]
Checking the ios Podfile.lock i realize there was no entry for the package, i don't know why is not auto linking it.

Adding `pod 'EXRandom', :path => '../node_modules/expo-random/ios'` into the Podfile fixes the issue

**Comment by santiripper** [comment]
Ok is not detecting the dependency because the file `unimodule.json` is missing.

The content should something like:

```json
{
  "name": "expo-random",
  "platforms": ["ios", "android", "web"]
}
```

**Comment by brentvatne** [comment]
@santiripper - i released expo-random@9.0.1 that satisfies the requirement of rn-cli autolinking to have a xcodeproj to be detected, so you should be able to update and then just run `npx pod-install` without adding anything to your podfile.

i'm going to also follow up on this to ensure that we work out the dependency situation with auth-session and random given that rn-cli autolinking (unlike expo autolinking) only works for direct dependencies, so expo-random needs to be in your project's package.json for now rather than installed via another package like expo-auth-session.
